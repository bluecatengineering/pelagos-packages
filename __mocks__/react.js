module.exports = {
	...jest.requireActual('react'),
	useState: jest.fn((v) => [typeof v === 'function' ? v() : v, jest.fn()]),
	useCallback: (fn) => fn,
	useContext: jest.fn(),
	useEffect: jest.fn(),
	useLayoutEffect: jest.fn(),
	useMemo: jest.fn((f) => f()),
	useReducer: jest.fn((reducer, initial) => [initial, jest.fn()]),
	useRef: jest.fn((current) => ({current})),
	Children: {
		toArray(children) {
			return Array.isArray(children) ? children : Array.of(children);
		},
		map(children, fn) {
			return this.toArray(children).map(fn);
		},
	},
};
