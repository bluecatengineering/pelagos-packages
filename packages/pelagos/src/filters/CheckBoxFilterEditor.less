@import '../../less/fonts';
@import '../../less/spacing';

.CheckBoxFilterEditor {
	border: none;
	display: flex;
	flex-direction: column;
	margin: 0;
	padding: 0;

	&__legend {
		@font-12-400();
		color: var(--text-secondary);
		margin: 0 0 @sp-08;
		padding: 0;
	}

	&__option {
		margin-bottom: @sp-04;
	}
}
