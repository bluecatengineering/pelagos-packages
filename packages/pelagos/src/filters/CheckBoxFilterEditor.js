import {useCallback, useState} from 'react';
import PropTypes from 'prop-types';

import CheckBox from '../components/CheckBox';

import FilterEditor from './FilterEditor';

import './CheckBoxFilterEditor.less';

/** Filter editor where options are presented as check boxes. */
export const OldCheckBoxFilterEditor = ({label, options, list, getLabel, onListChange}) => (
	<fieldset className="CheckBoxFilterEditor">
		<legend className="CheckBoxFilterEditor__legend">{label}</legend>
		{options.map((opt) => (
			<CheckBox
				key={opt}
				id={`checkBoxOption-${opt}`}
				className="CheckBoxFilterEditor__option"
				label={getLabel(opt)}
				checked={list && list.includes(opt)}
				onChange={() =>
					onListChange(
						list
							? list.includes(opt)
								? list.length === 1
									? null
									: list.filter((cur) => cur !== opt)
								: [...list, opt]
							: [opt]
					)
				}
			/>
		))}
	</fieldset>
);

OldCheckBoxFilterEditor.propTypes = {
	/** The field label. */
	label: PropTypes.string.isRequired,
	/** The filter options. */
	options: PropTypes.array.isRequired,
	/** The current filter values. */
	list: PropTypes.array,
	/** Function returning the option label. */
	getLabel: PropTypes.func.isRequired,
	/** Function invoked to apply changes. */
	onListChange: PropTypes.func,
};

/** Filter editor where options are presented as check boxes. */
export const NewCheckBoxFilterEditor = ({id, label, options, list: initialList, getLabel, chipId, onClose, onSave}) => {
	const [list, setList] = useState(initialList || []);
	const handleChange = useCallback(
		(event) => {
			const {checked, value} = event.target;
			const option = options[value];
			setList((list) => (checked ? [...list, option] : list.filter((cur) => cur !== option)));
		},
		[options]
	);
	const handleSave = useCallback(() => onSave(list.length ? list : null), [list, onSave]);
	return (
		<FilterEditor id={id} chipId={chipId} onClose={onClose} onSave={handleSave}>
			<fieldset className="CheckBoxFilterEditor">
				<legend className="CheckBoxFilterEditor__legend">{label}</legend>
				{options.map((opt, index) => (
					<CheckBox
						key={opt}
						id={`${id}-${opt}`}
						className="CheckBoxFilterEditor__option"
						label={getLabel(opt)}
						value={index}
						checked={list.includes(opt)}
						onChange={handleChange}
					/>
				))}
			</fieldset>
		</FilterEditor>
	);
};

NewCheckBoxFilterEditor.propTypes = {
	/** The component id. */
	id: PropTypes.string,
	/** The field label. */
	label: PropTypes.string.isRequired,
	/** The filter options. */
	options: PropTypes.array.isRequired,
	/** The current filter values. */
	list: PropTypes.array,
	/** Function returning the option label. */
	getLabel: PropTypes.func.isRequired,
	/** The ID of the chip which triggered the editor. */
	chipId: PropTypes.string,
	/** Function invoked to close the editor. */
	onClose: PropTypes.func,
	/** Function invoked to save changes. */
	onSave: PropTypes.func,
};

/** Filter editor where options are presented as check boxes. */
const CheckBoxFilterEditor = ({forArea, ...props}) =>
	forArea ? <NewCheckBoxFilterEditor {...props} /> : <OldCheckBoxFilterEditor {...props} />;

CheckBoxFilterEditor.propTypes = {
	/** The component id (only when forArea is true). */
	id: PropTypes.string,
	/** The field label. */
	label: PropTypes.string.isRequired,
	/** The filter options. */
	options: PropTypes.array.isRequired,
	/** The current filter values. */
	list: PropTypes.array,
	/** Function returning the option label. */
	getLabel: PropTypes.func.isRequired,
	/** Function invoked to apply changes (only when forArea is false). */
	onListChange: PropTypes.func,
	/** The ID of the chip which triggered the editor (only when forArea is true). */
	chipId: PropTypes.string,
	/** Function invoked to close the editor (only when forArea is true). */
	onClose: PropTypes.func,
	/** Function invoked to save changes (only when forArea is true). */
	onSave: PropTypes.func,
	/** Whether to enable new behaviour for FilterArea.  */
	forArea: PropTypes.bool,
};

export default CheckBoxFilterEditor;
