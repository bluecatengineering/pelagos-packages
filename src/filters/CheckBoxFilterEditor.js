import React from 'react';
import PropTypes from 'prop-types';

import Label from '../components/Label';
import CheckBox from '../components/CheckBox';

import './CheckBoxFilterEditor.less';

/** Filter editor where options are presented as check boxes. */
const CheckBoxFilterEditor = ({label, options, list, getLabel, onListChange}) => (
	<>
		<div className="CheckBoxFilterEditor__label">
			<Label text={label} />
		</div>
		{options.map((opt) => (
			<CheckBox
				key={opt}
				id={`checkBoxOption-${opt}`}
				className="CheckBoxFilterEditor__option"
				label={getLabel(opt)}
				checked={list && list.includes(opt)}
				onChange={() =>
					onListChange(
						list
							? list.includes(opt)
								? list.length === 1
									? null
									: list.filter((cur) => cur !== opt)
								: [...list, opt]
							: [opt]
					)
				}
			/>
		))}
	</>
);

CheckBoxFilterEditor.propTypes = {
	/** The field label. */
	label: PropTypes.string.isRequired,
	/** The filter options. */
	options: PropTypes.array.isRequired,
	/** The current filter values. */
	list: PropTypes.array,
	/** Function returning the option label. */
	getLabel: PropTypes.func.isRequired,
	/** Function invoked to apply changes. */
	onListChange: PropTypes.func,
};

export default CheckBoxFilterEditor;
