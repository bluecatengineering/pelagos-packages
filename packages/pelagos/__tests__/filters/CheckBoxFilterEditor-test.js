import {useState} from 'react';
import {shallow} from 'enzyme';

import CheckBoxFilterEditor, {
	NewCheckBoxFilterEditor,
	OldCheckBoxFilterEditor,
} from '../../src/filters/CheckBoxFilterEditor';

jest.unmock('../../src/filters/CheckBoxFilterEditor');

const anyFunction = expect.any(Function);

describe('CheckBoxFilterEditor', () => {
	describe('rendering', () => {
		it('renders expected elements when forArea is not set', () => {
			const wrapper = shallow(<CheckBoxFilterEditor label="label" options={[]} getLabel={jest.fn()} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when forArea is set', () => {
			const wrapper = shallow(<CheckBoxFilterEditor label="label" options={[]} getLabel={jest.fn()} forArea />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('OldCheckBoxFilterEditor', () => {
		describe('rendering', () => {
			it('renders expected elements', () => {
				const wrapper = shallow(
					<OldCheckBoxFilterEditor label="label" list={['a']} options={['a', 'b']} getLabel={(o) => o} />
				);
				expect(wrapper.getElement()).toMatchSnapshot();
			});
		});

		describe('behaviour', () => {
			it('adds an item to the list', () => {
				const onListChange = jest.fn();
				const wrapper = shallow(
					<OldCheckBoxFilterEditor
						label="label"
						list={['b']}
						options={['a']}
						getLabel={jest.fn()}
						onListChange={onListChange}
					/>
				);
				wrapper.find('#checkBoxOption-a').prop('onChange')();
				expect(onListChange.mock.calls).toEqual([[['b', 'a']]]);
			});

			it('adds an item to the list when the list is null', () => {
				const onListChange = jest.fn();
				const wrapper = shallow(
					<OldCheckBoxFilterEditor
						label="label"
						list={null}
						options={['a']}
						getLabel={jest.fn()}
						onListChange={onListChange}
					/>
				);
				wrapper.find('#checkBoxOption-a').prop('onChange')();
				expect(onListChange.mock.calls).toEqual([[['a']]]);
			});

			it('removes an item from the list', () => {
				const onListChange = jest.fn();
				const wrapper = shallow(
					<OldCheckBoxFilterEditor
						label="label"
						list={['a', 'b']}
						options={['a']}
						getLabel={jest.fn()}
						onListChange={onListChange}
					/>
				);
				wrapper.find('#checkBoxOption-a').prop('onChange')();
				expect(onListChange.mock.calls).toEqual([[['b']]]);
			});

			it('removes the last item from the list', () => {
				const onListChange = jest.fn();
				const wrapper = shallow(
					<OldCheckBoxFilterEditor
						label="label"
						list={['a']}
						options={['a']}
						getLabel={jest.fn()}
						onListChange={onListChange}
					/>
				);
				wrapper.find('#checkBoxOption-a').prop('onChange')();
				expect(onListChange.mock.calls).toEqual([[null]]);
			});
		});
	});

	describe('NewCheckBoxFilterEditor', () => {
		describe('rendering', () => {
			it('renders expected elements', () => {
				const wrapper = shallow(
					<NewCheckBoxFilterEditor
						id="test"
						label="label"
						list={['a']}
						options={['a', 'b']}
						getLabel={(o) => o}
						chipId="button"
					/>
				);
				expect(wrapper.getElement()).toMatchSnapshot();
			});

			it('renders expected elements when list is not set', () => {
				const wrapper = shallow(
					<NewCheckBoxFilterEditor id="test" label="label" options={['a', 'b']} getLabel={(o) => o} chipId="button" />
				);
				expect(wrapper.getElement()).toMatchSnapshot();
			});
		});

		describe('behaviour', () => {
			it('adds an item to the list', () => {
				const list = ['b'];
				const setList = jest.fn();
				useState.mockReturnValueOnce([list, setList]);
				const wrapper = shallow(
					<NewCheckBoxFilterEditor id="test" label="label" list={list} options={['a']} getLabel={jest.fn()} />
				);
				wrapper.find('#test-a').simulate('change', {target: {checked: true, value: '0'}});
				expect(setList.mock.calls).toEqual([[anyFunction]]);
				expect(setList.mock.calls[0][0](list)).toEqual(['b', 'a']);
			});

			it('removes an item from the list', () => {
				const list = ['a', 'b'];
				const setList = jest.fn();
				useState.mockReturnValueOnce([list, setList]);
				const wrapper = shallow(
					<NewCheckBoxFilterEditor id="test" label="label" list={list} options={['a']} getLabel={jest.fn()} />
				);
				wrapper.find('#test-a').simulate('change', {target: {checked: false, value: '0'}});
				expect(setList.mock.calls).toEqual([[anyFunction]]);
				expect(setList.mock.calls[0][0](list)).toEqual(['b']);
			});

			it('calls onSave when the editor calls onSave', () => {
				const list = ['a'];
				const onSave = jest.fn();
				const wrapper = shallow(
					<NewCheckBoxFilterEditor
						id="test"
						label="label"
						list={list}
						options={['a', 'b']}
						getLabel={(o) => o}
						onSave={onSave}
					/>
				);
				wrapper.find('#test').prop('onSave')();
				expect(onSave.mock.calls).toEqual([[list]]);
			});

			it('calls onSave when the editor calls onSave and the list is empty', () => {
				const onSave = jest.fn();
				const wrapper = shallow(
					<NewCheckBoxFilterEditor
						id="test"
						label="label"
						list={[]}
						options={['a', 'b']}
						getLabel={(o) => o}
						onSave={onSave}
					/>
				);
				wrapper.find('#test').prop('onSave')();
				expect(onSave.mock.calls).toEqual([[null]]);
			});
		});
	});
});
