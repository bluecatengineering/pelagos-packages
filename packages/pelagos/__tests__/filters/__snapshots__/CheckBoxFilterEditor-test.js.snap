// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckBoxFilterEditor rendering renders expected elements 1`] = `
[
  <fieldset
    className="CheckBoxFilterEditor"
  >
    <legend
      className="CheckBoxFilterEditor__legend"
    >
      label
    </legend>
    <CheckBox
      checked={true}
      className="CheckBoxFilterEditor__option"
      id="checkBoxOption-a"
      label="a"
      onChange={[Function]}
    />
    <CheckBox
      checked={false}
      className="CheckBoxFilterEditor__option"
      id="checkBoxOption-b"
      label="b"
      onChange={[Function]}
    />
  </fieldset>,
]
`;
